File: src/main/java/no/nav/permitteringsskjemaapi/exceptions/PermitteringExceptionHandler.java
package no.nav.permitteringsskjemaapi.exceptions;

import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import no.nav.permitteringsskjemaapi.util.TokenUtil;
import no.nav.security.token.support.core.exceptions.JwtTokenValidatorException;
import no.nav.security.token.support.spring.validation.interceptor.JwtTokenUnauthorizedException;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.client.HttpStatusCodeException;
import org.springframework.web.context.request.WebRequest;
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;

import java.util.List;
import java.util.Optional;

import static java.util.Arrays.asList;
import static java.util.stream.Collectors.toList;
import static org.springframework.http.HttpStatus.*;

@ControllerAdvice
@AllArgsConstructor
@Slf4j
public class PermitteringExceptionHandler extends ResponseEntityExceptionHandler {
    private final TokenUtil tokenUtil;

    @ResponseBody
    @ExceptionHandler(HttpStatusCodeException.class)
    public ResponseEntity<Object> handleHttpStatusCodeException(HttpStatusCodeException e, WebRequest request) {
        if (e.getStatusCode().equals(UNAUTHORIZED) || e.getStatusCode().equals(FORBIDDEN)) {
            return logAndHandle(e.getStatusCode(), e, request, tokenUtil.getExpiryDate());
        }
        return logAndHandle(e.getStatusCode(), e, request);
    }

    @Override
    protected ResponseEntity<Object> handleMethodArgumentNotValid(MethodArgumentNotValidException e,
                                                                  HttpHeaders headers, HttpStatus status, WebRequest req) {
        return logAndHandle(UNPROCESSABLE_ENTITY, e, req, headers, validationErrors(e));
    }

    @Override
    protected ResponseEntity<Object> handleHttpMessageNotReadable(HttpMessageNotReadableException e,
                                                                  HttpHeaders headers, HttpStatus status, WebRequest req) {
        return logAndHandle(UNPROCESSABLE_ENTITY, e, req, headers);
    }

    @ExceptionHandler({ UnexpectedInputException.class })
    public ResponseEntity<Object> handleUnexpectedException(UnexpectedInputException e, WebRequest req) {
        return logAndHandle(UNPROCESSABLE_ENTITY, e, req);
    }

    @ExceptionHandler({ PermitteringsApiException.class })
    public ResponseEntity<Object> apiException(PermitteringsApiException e, WebRequest req) {
        return logAndHandle(UNPROCESSABLE_ENTITY, e, req);
    }

    @ExceptionHandler({ JwtTokenUnauthorizedException.class })
    public ResponseEntity<Object> handleUnauthorizedException(JwtTokenUnauthorizedException e, WebRequest req) {
        return logAndHandle(UNAUTHORIZED, e, req);
    }

    @ExceptionHandler({ JwtTokenValidatorException.class })
    public ResponseEntity<Object> handleUnauthenticatedJwtException(JwtTokenValidatorException e, WebRequest req) {
        return logAndHandle(FORBIDDEN, e, req, e.getExpiryDate());
    }

    @ExceptionHandler(value = { Exception.class })
    protected ResponseEntity<Object> handleUncaught(Exception e, WebRequest req) {
        return logAndHandle(INTERNAL_SERVER_ERROR, e, req);
    }

    @ExceptionHandler(value = { IkkeTilgangException.class })
    protected ResponseEntity<Object> ikkeTilgang(Exception e, WebRequest req) {
        return logAndHandle(UNAUTHORIZED, e, req);
    }

    @ExceptionHandler(value = { IkkeFunnetException.class })
    protected ResponseEntity<Object> ikkeFunnet(Exception e, WebRequest req) {
        return logAndHandle(NOT_FOUND, e, req);
    }

    @ExceptionHandler(value = { AlleFelterIkkeFyltUtException.class })
    protected ResponseEntity<Object> alleFelterIkkeFyltUt(Exception e, WebRequest req) {
        return logAndHandle(BAD_REQUEST, e, req);
    }

    @ExceptionHandler(value = { SkjemaErAvbruttException.class })
    protected ResponseEntity<Object> avbrutt(Exception e, WebRequest req) {
        return logAndHandle(BAD_REQUEST, e, req);
    }

    private ResponseEntity<Object> logAndHandle(HttpStatus status, Exception e, WebRequest req, Object... messages) {
        return logAndHandle(status, e, req, new HttpHeaders(), messages);
    }

    private ResponseEntity<Object> logAndHandle(HttpStatus status, Exception e, WebRequest req, HttpHeaders headers,
                                                Object... messages) {
        return logAndHandle(status, e, req, headers, asList(messages));
    }

    private ResponseEntity<Object> logAndHandle(HttpStatus status, Exception e, WebRequest req, HttpHeaders headers,
                                                List<Object> messages) {
        ApiError apiError = apiErrorFra(status, e, messages);
        log.warn("({}) {} {} ({})", subject(), status, apiError.getMessages(), status.value(), e);
        return handleExceptionInternal(e, apiError, headers, status, req);
    }

    private String subject() {
        return Optional.ofNullable(tokenUtil.getSubject())
                .orElse("Uautentisert");
    }

    private static ApiError apiErrorFra(HttpStatus status, Exception e, List<Object> messages) {
        return new ApiError(status, e, messages);
    }

    private static List<String> validationErrors(MethodArgumentNotValidException e) {
        return e.getBindingResult().getFieldErrors()
                .stream()
                .map(PermitteringExceptionHandler::errorMessage)
                .collect(toList());
    }

    private static String errorMessage(FieldError error) {
        return error.getField() + " " + error.getDefaultMessage();
    }
}

File: src/main/kotlin/no/nav/permitteringsskjemaapi/exceptions/PermitteringExceptionHandler.kt
package no.nav.permitteringsskjemaapi.exceptions

import com.fasterxml.jackson.core.JsonProcessingException
import no.nav.permitteringsskjemaapi.config.logger
import no.nav.permitteringsskjemaapi.util.TokenUtil
import no.nav.security.token.support.core.exceptions.JwtTokenValidatorException
import no.nav.security.token.support.spring.validation.interceptor.JwtTokenUnauthorizedException
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.HttpStatusCode
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.validation.FieldError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseBody
import org.springframework.web.client.HttpStatusCodeException
import org.springframework.web.context.request.WebRequest
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler
import java.util.stream.Collectors

@ControllerAdvice
class PermitteringExceptionHandler(private val tokenUtil: TokenUtil) : ResponseEntityExceptionHandler() {
    private val log = logger()

    @ResponseBody
    @ExceptionHandler(
        HttpStatusCodeException::class
    )
    fun handleHttpStatusCodeException(e: HttpStatusCodeException, request: WebRequest): ResponseEntity<Any>? {
        return if (e.statusCode.isSameCodeAs(HttpStatus.UNAUTHORIZED) || e.statusCode.isSameCodeAs(HttpStatus.FORBIDDEN)) {
            logAndHandle(e.statusCode, e, request, messages = listOf(tokenUtil.expiryDate))
        } else
            logAndHandle(e.statusCode, e, request, messages = listOf<Any>())
    }

    override fun handleMethodArgumentNotValid(
        e: MethodArgumentNotValidException,
        headers: HttpHeaders,
        status: HttpStatusCode,
        req: WebRequest
    ): ResponseEntity<Any>? {
        return logAndHandle(HttpStatus.UNPROCESSABLE_ENTITY, e, req, headers, validationErrors(e))
    }

    override fun handleHttpMessageNotReadable(
        e: HttpMessageNotReadableException,
        headers: HttpHeaders,
        status: HttpStatusCode,
        req: WebRequest
    ): ResponseEntity<Any>? {
        return logAndHandle(HttpStatus.UNPROCESSABLE_ENTITY, e, req, headers)
    }

    @ExceptionHandler(JsonProcessingException::class)
    fun handleJsonProcessingException(e: JsonProcessingException, req: WebRequest): ResponseEntity<Any>? {
        return logAndHandle(HttpStatus.UNPROCESSABLE_ENTITY, e, req)
    }

    @ExceptionHandler(PermitteringsApiException::class)
    fun apiException(e: PermitteringsApiException, req: WebRequest): ResponseEntity<Any>? {
        return logAndHandle(HttpStatus.UNPROCESSABLE_ENTITY, e, req)
    }

    @ExceptionHandler(JwtTokenUnauthorizedException::class)
    fun handleUnauthorizedException(e: JwtTokenUnauthorizedException, req: WebRequest): ResponseEntity<Any>? {
        return logAndHandle(HttpStatus.UNAUTHORIZED, e, req)
    }

    @ExceptionHandler(JwtTokenValidatorException::class)
    fun handleUnauthenticatedJwtException(e: JwtTokenValidatorException, req: WebRequest): ResponseEntity<Any>? {
        return logAndHandle(HttpStatus.FORBIDDEN, e, req, messages = listOf(e.expiryDate))
    }

    @ExceptionHandler(value = [Exception::class])
    protected fun handleUncaught(e: Exception, req: WebRequest): ResponseEntity<Any>? {
        return logAndHandle(HttpStatus.INTERNAL_SERVER_ERROR, e, req)
    }

    @ExceptionHandler(value = [IkkeTilgangException::class])
    protected fun ikkeTilgang(e: Exception, req: WebRequest): ResponseEntity<Any>? {
        return logAndHandle(HttpStatus.FORBIDDEN, e, req)
    }

    @ExceptionHandler(value = [IkkeFunnetException::class])
    protected fun ikkeFunnet(e: Exception, req: WebRequest): ResponseEntity<Any>? {
        return logAndHandle(HttpStatus.NOT_FOUND, e, req)
    }

    @ExceptionHandler(value = [AlleFelterIkkeFyltUtException::class])
    protected fun alleFelterIkkeFyltUt(e: Exception, req: WebRequest): ResponseEntity<Any>? {
        return logAndHandle(HttpStatus.BAD_REQUEST, e, req)
    }

    @ExceptionHandler(value = [SkjemaErAvbruttException::class])
    protected fun avbrutt(e: Exception, req: WebRequest): ResponseEntity<Any>? {
        return logAndHandle(HttpStatus.BAD_REQUEST, e, req)
    }

    private fun logAndHandle(
        status: HttpStatusCode,
        e: Exception,
        req: WebRequest,
        headers: HttpHeaders = HttpHeaders(),
        messages: List<Any?> = emptyList(),
    ): ResponseEntity<Any>? {
        val apiError = ApiError(status, e, messages)
        log.warn("{} {} ({})", status, apiError.messages, status.value(), e)
        return handleExceptionInternal(e, apiError, headers, status, req)
    }

    companion object {
        private fun validationErrors(e: MethodArgumentNotValidException): List<String> {
            return e.bindingResult.fieldErrors
                .stream()
                .map { error: FieldError -> errorMessage(error) }
                .collect(Collectors.toList())
        }

        private fun errorMessage(error: FieldError): String {
            return error.field + " " + error.defaultMessage
        }
    }
}